polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -47,38 +47,112 @@ namespace polygeist {
 
-namespace {
+/// Returns true if the callable operation \p callableOp has a callable region
+/// and private visibility, and false otherwise.
+bool isPrivateCallable(CallableOpInterface callableOp) {
+  auto sym = dyn_cast<SymbolOpInterface>(callableOp.getOperation());
+  return (sym && sym.isPrivate() && callableOp.getCallableRegion());
+}
 
-/// Represent a candidate for the argument promotion transformation.
-class Candidate {
+/// Returns true if the call \p callOp is recursive, and false otherwise.
+bool isRecursiveCall(CallOpInterface callOp) {
+  auto callable = dyn_cast<CallableOpInterface>(callOp.resolveCallable());
+  return (callable && callable.getCallableRegion() &&
+          callable.getCallableRegion()->isAncestor(callOp->getParentRegion()));
+}
+
+/// Returns true if \p type is 'memref<?xstruct<>>', and false otherwise.
+bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  auto structType = dyn_cast<LLVM::LLVMStructType>(mt.getElementType());
+  if (!structType || llvm::any_of(structType.getBody(), [](Type memType) {
+        return isa<LLVM::LLVMStructType>(memType);
+      }))
+    return false;
+  return true;
+}
+
+// Returns true if the block of \p startOp contains an instruction (after \p
+// startOp) that has a side effects on \p val.
+auto existsSideEffectAfter(Value val, Operation *startOp) {
+  assert(startOp && "Expecting a valid pointer");
+  Block *block = startOp->getBlock();
+
+  for (Operation &op : *block) {
+    if (op.isBeforeInBlock(startOp) || isMemoryEffectFree(&op))
+      continue;
+
+    // Operations with unknown side effects might write to any memory.
+    if (isa<MemoryEffectOpInterface>(op) &&
+        op.hasTrait<OpTrait::HasRecursiveMemoryEffects>())
+      continue;
+
+    // Currently bail out if any subsequent operation has side effects.instruc
+    // TODO: check whether the operation alias with 'val'.
+    if (auto MEI = dyn_cast<MemoryEffectOpInterface>(op))
+      return true;
+  }
+  return false;
+}
+
+/// Represents an operand that can be peeled.
+class CandidateOperand {
 public:
-  Candidate(CallOpInterface callOp) : callOp(callOp) {
-    assert(!isRecursiveCall(callOp) && "Callee must not be recursive");
+  CandidateOperand(Value v, unsigned pos) : val(v), position(pos) {
+    assert(isValidMemRefType(val.getType()) &&
+           "Candidate operand does not have the expected type");
   }
 
-  /// Returns true if the callable operation \p callableOp has a callable region
-  /// and private visibility, and false otherwise.
-  static bool isPrivateCallable(CallableOpInterface callableOp);
+  Value value() const { return val; };
+  unsigned pos() const { return position; };
```

Comment:
optional, this is a personal preference.
```suggestion
  unsigned position() const { return position; };
```

Reply:
Agree, changed.